### DISCUSSION TITLE
Apply element enable/disable to on_click/on_change events

### DISCUSSION THREAD

**Discussion by thetableman** [zauberzeug/nicegui#983] [Ideas / Feature Requests]
When an element is enabled or disabled this affects the web ui only, it's fairly easy for a user to overcome this by adjusting the html using a browsers inspect tool.

Obviously it's best practice to follow up a ui disable with some back-end validation, however, it would be useful if the enable/disable affected whether a on_click or on_change fired. This could be an additional method or parameter.

For example:
```py
button = ui.button('Click Me', on_click=lambda: ui.notify('Clicked'))
button.disable(disable_events=True) #could default to False so as not to impact current practice
button.set_visibility(False, disable_events=False)
```
or
```py
button = ui.button('Click Me', on_click=lambda: ui.notify('Clicked'), on_disable_events=True, on_visibility_events=True)
button.disable() #would disable on_click also
button.set_visibility(False) 
```

This would reduce code to validate after a disable in simple cases and truly 'disable' an element.

**Comment by falkoschindler** [comment]
Thanks for the proposal, @thetableman!
We are, however, not convinced that this kind of check needs to be implemented in NiceGUI. It increases code and API complexity quite a bit without any advantages for the "normal" user. If security against malicious attacks is an issue, you might want to add certain checks in the user code.
But maybe there are other opinions on this matter?

**Reply by thetableman** [reply]
@falkoschindler,  it's not that I'm particularly concerned about security or malicious attacks, it's just best practice to apply additional controls to ensure a user can't execute a process or access a page by working around changes to UI elements like visibility and disable. Any user who plans to utilise NiceGUI in production with public access would be foolish to rely purely on hiding a button to prevent a form being executed or a page being accessible, they would be wise to follow up with extra logic to ensure the associated function doesn't run at all.

So I do respectfully disagree - a "normal" user would definitely find disabling a button/toggle/etc callback useful. Regardless, I appreciate your consideration of my suggestion and understand that not all ideas can be implemented.

**Reply by falkoschindler** [reply]
@thetableman You're starting to convince us! 😉

I talked to ChatGPT for a second opinion, and it absolutely confirms your point. If we find a neat place to implement this server-side validation, we might improve the overall robustness, integrity and security of NiceGUI applications.

We should probably activate the validation by default and allow the user to opt-out if needed. I could imaging to add a check somewhere central like in `handle_event` in events.py. If the sender is a `DisableableElement` it can read something like an `ignore_events_when_disabled` attribute. And if it derives from `Visibility` (every `Element` does), there could be an attribute `ignore_events_when_invisible`. This way there would be no need for additional initializer or method parameters.

**Reply by thetableman** [reply]
Thanks, @falkoschindler. I think this would be a really useful addition that cuts back doubling-up on validation code.

**Comment by falkoschindler** [comment]
Just for reference, here is an example how one can inject a click event on a disabled button:

```py
b = ui.button('Click me', on_click=lambda: ui.label('You did it!'))
ui.switch('Enable').bind_value_to(b, 'enabled')
ui.button('Inject click', on_click=lambda: ui.run_javascript(f'getElement({b.id}).$emit("click")', respond=False))
```

**Comment by falkoschindler** [comment]
I just created PR https://github.com/zauberzeug/nicegui/pull/1030. What do you think, @thetableman?

**Reply by falkoschindler** [reply]
Has been released in version 1.2.21.

**Reply by jabonita-arch** [reply]
how to do it?? im new

**Reply by falkoschindler** [reply]
Hi @jabonita-arch, welcome! The good news: there is nothing you need to do. Since version 1.2.21, NiceGUI ignores events from disabled and hidden elements on the server side by default. So if you disable a button, its click handler won't run even if someone re-enables the button via the browser's dev tools:

```py
from nicegui import ui

button = ui.button('Click me', on_click=lambda: ui.notify('Clicked'))
button.disable()

ui.run()
```

The same applies to `button.set_visibility(False)`.

If you ever need the old behavior (handlers running even while the element is disabled or hidden), you can opt out per element:

```py
button.ignores_events_when_disabled = False
button.ignores_events_when_hidden = False
```

If you meant something else, feel free to open a new discussion with a small example of what you're trying to build. [※](https://zauberzeug.github.io/colophon/#m=claude-fable-5-1&t=Drafted%20by%20the%20model%20from%20the%20discussion%20history%20and%20the%20current%20NiceGUI%20source%2C%20at%20the%20maintainer%27s%20request.&a=claude-code "claude-fable-5-1: Drafted by the model from the discussion history and the current NiceGUI source, at the maintainer's request.")
